record_batch.go

```diff
@@ -0,0 +1,204 @@
+package kafka
+
+import (
+	"bufio"
+	"bytes"
+	"hash/crc32"
+	"time"
+)
+
+type record struct {
+	attributes int8
+	baseTime   time.Time
+	offset     int64
+	msg        Message
+}
+
+func (r record) size() int32 {
+	timestampDelta := r.msg.Time.Sub(r.baseTime)
+	offsetDelta := int64(r.offset)
+
+	var res int
+	res += 1 + // attributes
+		varIntLen(int64(milliseconds(timestampDelta))) +
+		varIntLen(offsetDelta) +
+		varIntLen(int64(len(r.msg.Key))) +
+		len(r.msg.Key) +
+		varIntLen(int64(len(r.msg.Value))) +
+		len(r.msg.Value) +
+		varIntLen(int64(len(r.msg.Headers)))
+	for _, h := range r.msg.Headers {
+		res += varIntLen(int64(len([]byte(h.Key)))) +
+			len([]byte(h.Key)) +
+			varIntLen(int64(len(h.Value))) +
+			len(h.Value)
+	}
+	return int32(res)
+}
+
+func (r record) writeTo(w *bufio.Writer) {
+	timestampDelta := r.msg.Time.Sub(r.baseTime)
+	offsetDelta := int64(r.offset)
+
+	writeVarInt(w, int64(r.size()))
+
+	writeInt8(w, r.attributes)
+	writeVarInt(w, int64(milliseconds(timestampDelta)))
+	writeVarInt(w, offsetDelta)
+
+	writeVarInt(w, int64(len(r.msg.Key)))
+	w.Write(r.msg.Key)
+	writeVarInt(w, int64(len(r.msg.Value)))
+	w.Write(r.msg.Value)
+	writeVarInt(w, int64(len(r.msg.Headers)))
+
+	for _, h := range r.msg.Headers {
+		writeVarInt(w, int64(len(h.Key)))
+		w.Write([]byte(h.Key))
+		writeVarInt(w, int64(len(h.Value)))
+		w.Write(h.Value)
+	}
+}
+
+type recordBatch struct {
+	codec      CompressionCodec
+	attributes int16
+	msgs       []Message
+	compressed []byte
+}
+
+func (r *recordBatch) Init() (err error) {
```

Comment:
```suggestion
func (r *recordBatch) init() (err error) {
```
Unless I'm mistaking, it seems like this method is only used in the package, there's no need for it to be exported.